compiler/ast.rs

```diff
@@ -0,0 +1,633 @@
+      }),
+      helpers::InjectHelpers,
+      typescript::strip(),
+      fixer(Some(&self.comments)),
+    );
+    let program = swc_common::GLOBALS.set(&Globals::new(), || {
+      helpers::HELPERS.set(&helpers::Helpers::new(false), || {
+        if transform_jsx {
+          program.fold_with(&mut chain!(jsx_pass, compiler_pass))
+        } else {
+          program.fold_with(&mut compiler_pass)
+        }
+      })
+    });
+
+    let mut mappings = Vec::new();
+    let mut buf = Vec::new();
+    {
+      let handlers = Box::new(EmptyHandler);
+      let wr = Box::new(JsWriter::new(
+        self.source_map.clone(),
+        "\n",
+        &mut buf,
+        Some(&mut mappings),
+      ));
+      let cfg = Config { minify: false };
+      let mut emitter = swc_ecmascript::codegen::Emitter {
+        cfg,
+        comments: Some(&self.comments),
+        cm: self.source_map.clone(),
+        wr,
+        handlers,
+      };
+      program.emit_with(&mut emitter)?;
+    }
+    let mut code = String::from_utf8(buf)?;
+    {
+      let mut buf = Vec::new();
+      self
+        .source_map
+        .build_source_map_from(&mut mappings, None)
+        .to_writer(&mut buf)?;
+      let map = String::from_utf8(buf)?;
+
+      code.push_str("//# sourceMappingURL=data:application/json;base64,");
+      let encoded_map = base64::encode(map.as_bytes());
+      code.push_str(&encoded_map);
+    }
+
+    Ok(code)
+  }
+}
+
+/// For a given specifier, source, and media type, parse the source of the
+/// module and return a representation which can be further processed.
+pub fn parse(
+  specifier: &ModuleSpecifier,
+  source: &str,
+  media_type: MediaType,
+) -> Result<ParsedModule> {
+  let source_map = SourceMap::default();
+  let source_file = source_map.new_source_file(
+    FileName::Custom(specifier.to_string()),
+    source.to_string(),
+  );
+  let error_buffer = ErrorBuffer::new();
+  let syntax = get_syntax(media_type);
+  let input = StringInput::from(&*source_file);
+  let comments = SingleThreadedComments::default();
+
+  let handler = Handler::with_emitter_and_flags(
+    Box::new(error_buffer.clone()),
+    HandlerFlags {
+      can_emit_warnings: true,
+      dont_buffer_diagnostics: true,
+      ..HandlerFlags::default()
+    },
+  );
+
+  let lexer = Lexer::new(syntax, TARGET, input, Some(&comments));
+  let mut parser = swc_ecmascript::parser::Parser::new_from(lexer);
+
+  let sm = &source_map;
+  let module = parser.parse_module().map_err(move |err| {
+    let mut diagnostic = err.into_diagnostic(&handler);
+    diagnostic.emit();
+
+    ErrBox::from(DiagnosticBuffer::from_error_buffer(error_buffer, |span| {
+      sm.lookup_char_pos(span.lo)
+    }))
+  })?;
+
+  Ok(ParsedModule {
+    module,
+    source_map: Rc::new(source_map),
+    comments,
+  })
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DependencyDescriptor {
+  pub is_dynamic: bool,
+  pub is_type: bool,
+  pub span: Span,
```

Comment:
I suggest to use `location: Location` instead of span, as well as to extract leading comments during `DependencyCollector` run - it should limit APIs needed on `ParsedModule`